File: C-structless/game.c
#include "game.h"
#include "player.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef char Question[255];

Question pop_questions[50];
Question science_questions[50];
Question sports_questions[50];
Question rock_questions[50];

Question *pop_question;
Question *science_question;
Question *sports_question;
Question *rock_question;
extern  char *players[];
bool is_getting_out_of_penalty_box;
extern int not_a_winner;
static void create_rock_question ();
const char *current_category ( void );


void newGame ( void )
{
  int i;
  player_num = 0;
  current_player = 0;

  pop_question = pop_questions;
  science_question = science_questions;
  sports_question = sports_questions;
  rock_question = rock_questions;

  for (i = 0; i < 50; i++)
    {
      sprintf (pop_questions[i], "Pop Question %d", i);
      sprintf (science_questions[i], "Science Question %d", i);
      sprintf (sports_questions[i], "Sports Question %d", i);
      create_rock_question (i);
    }
}

extern int how_many_players (void);

void
create_rock_question (int index)
{
  sprintf (rock_questions[index], "Rock Question %d", index);
}

bool
is_playable ( void )
{
  return (how_many_players () >= 2);
}




void
ask_question ( void )
{
  if (!strcmp (current_category (), "Pop"))
    {
      printf ("%s\n", *(++pop_question));
    }
  if (!strcmp (current_category (), "Science"))
    {
      printf ("%s\n", *(++science_question));
    }
  if (!strcmp (current_category (), "Sports"))
    {
      printf ("%s\n", *(++sports_question));
    }
  if (!strcmp (current_category (), "Rock"))
    {
      printf ("%s\n", *(++rock_question));
    }
}

extern int places[];

const char *
current_category ( void )
{
  if (places[current_player] == 0)
    return "Pop";
  if (places[current_player] == 4)
    return "Pop";
  if (places[current_player] == 8)
    return "Pop";
  if (places[current_player] == 1)
    return "Science";
  if (places[current_player] == 5)
    return "Science";
  if (places[current_player] == 9)
    return "Science";
  if (places[current_player] == 2)
    return "Sports";
  if (places[current_player] == 6)
    return "Sports";
  if (places[current_player] == 10)
    return "Sports";
  return "Rock";
}

extern int purses[6];
void
was_correctly_answered (void )
{
  if (in_penalty_box[current_player])
    {
      if (is_getting_out_of_penalty_box)
	{
	  printf ("Answer was correct!!!!\n");
	  purses[current_player]++;
	  printf ("%s now has %d Gold Coins.\n",
		  players[current_player],
		  purses[current_player]);
	  bool winner = did_player_win ();
	  current_player++;
	  if (current_player == player_num)
	    current_player = 0;

	  not_a_winner = winner;
	}
      else
	{
	  current_player++;
	  if (current_player == player_num)
	    current_player = 0;
	  not_a_winner = true;
	}



    }
  else
    {

      printf ("Answer was corrent!!!!\n");
      purses[current_player]++;
      printf ("%s now has %d Gold Coins.\n",
	      players[current_player],
	      purses[current_player]);

      bool winner = did_player_win ();
      current_player++;
      if (current_player == player_num)
	current_player = 0;
    not_a_winner = winner;
    }
}

void
wrong_answer (void )
{
  printf ("Question was incorrectly answered\n");
  printf ("%s was sent to the penalty box\n",
	  players[current_player]);
  in_penalty_box[current_player] = true;

  current_player++;
  if (current_player == player_num)
    current_player = 0;
}

File: java/src/main/java/com/adaptionsoft/games/uglytrivia/Game.java
package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;
import java.util.LinkedList;

public class Game {

    private static final int QUESTIONS_PER_CATEGORY = 50;

    private ArrayList<String> players = new ArrayList<>();
    private int[] places = new int[6];
    private int[] purses  = new int[6];
    private boolean[] inPenaltyBox  = new boolean[6];
    
    private LinkedList<String> popQuestions = new LinkedList<>();
    private LinkedList<String> scienceQuestions = new LinkedList<>();
    private LinkedList<String> sportsQuestions = new LinkedList<>();
    private LinkedList<String> rockQuestions = new LinkedList<>();
    
    private int currentPlayer = 0;
    private boolean isGettingOutOfPenaltyBox;

    enum Category {
        POP("Pop", 0),
        SCIENCE("Science", 1),
        SPORTS("Sports", 2),
        ROCK("Rock", 3);

        private String description;
        private int position;

        Category(String description, int position) {
            this.description = description;
            this.position = position;
        }

        public static String of(int position) {
            int wrapAround = position % numberOfCategories();
            for (Category category: Category.values()) {
                if (category.position == wrapAround) {
                    return category.description;
                }
            }
            return "";
        }

        private static int numberOfCategories() {
            return Category.values().length;
        }
    }

    public Game() {
    	for (int i = 0; i < QUESTIONS_PER_CATEGORY; i++) {
			popQuestions.addLast(questionFor(Category.POP, i));
			scienceQuestions.addLast(questionFor(Category.SCIENCE, i));
			sportsQuestions.addLast(questionFor(Category.SPORTS, i));
			rockQuestions.addLast(questionFor(Category.ROCK, i));
    	}
    }

	public String questionFor(Category category, int index) {
        return String.format(
                "%s Question %d",
                category.description,
                index);
	}

    public boolean add(String playerName) {

	    players.add(playerName);
	    places[howManyPlayers()] = 0;
	    purses[howManyPlayers()] = 0;
	    inPenaltyBox[howManyPlayers()] = false;

	    System.out.println(playerName + " was added");
	    System.out.println("They are player number " + players.size());
		return true;
	}

	public int howManyPlayers() {
		return players.size();
	}

	public void roll(int roll) {
		System.out.println(players.get(currentPlayer) + " is the current player");
		System.out.println("They have rolled a " + roll);

		if (inPenaltyBox[currentPlayer] && (roll % 2 == 0)) {
            System.out.println(players.get(currentPlayer) + " is not getting out of the penalty box");
            isGettingOutOfPenaltyBox = false;
            return;
        }

        if (inPenaltyBox[currentPlayer]) {
            isGettingOutOfPenaltyBox = true;
            System.out.println(players.get(currentPlayer) + " is getting out of the penalty box");
        }

        places[currentPlayer] = places[currentPlayer] + roll;
        if (places[currentPlayer] > 11) places[currentPlayer] = places[currentPlayer] - 12;

        System.out.println(players.get(currentPlayer)
                + "'s new location is "
                + places[currentPlayer]);
        System.out.println("The category is " + currentCategory());
        askQuestion();

	}

	private void askQuestion() {
		if (currentCategory().equals("Pop"))
			System.out.println(popQuestions.removeFirst());
		if (currentCategory().equals("Science"))
			System.out.println(scienceQuestions.removeFirst());
		if (currentCategory().equals("Sports"))
			System.out.println(sportsQuestions.removeFirst());
		if (currentCategory().equals("Rock"))
			System.out.println(rockQuestions.removeFirst());
	}


	private String currentCategory() {
        return Category.of(places[currentPlayer]);
	}

	public boolean wasCorrectlyAnswered() {
		if (inPenaltyBox[currentPlayer]){
			if (isGettingOutOfPenaltyBox) {
				System.out.println("Answer was correct!!!!");
				purses[currentPlayer]++;
				System.out.println(players.get(currentPlayer) 
						+ " now has "
						+ purses[currentPlayer]
						+ " Gold Coins.");
				
				boolean winner = didPlayerWin();
				currentPlayer++;
				if (currentPlayer == players.size()) currentPlayer = 0;
				
				return winner;
			} else {
				currentPlayer++;
				if (currentPlayer == players.size()) currentPlayer = 0;
				return true;
			}
			
			
			
		} else {
		
			System.out.println("Answer was corrent!!!!");
			purses[currentPlayer]++;
			System.out.println(players.get(currentPlayer) 
					+ " now has "
					+ purses[currentPlayer]
					+ " Gold Coins.");
			
			boolean winner = didPlayerWin();
			currentPlayer++;
			if (currentPlayer == players.size()) currentPlayer = 0;
			
			return winner;
		}
	}
	
	public boolean wrongAnswer(){
		System.out.println("Question was incorrectly answered");
		System.out.println(players.get(currentPlayer)+ " was sent to the penalty box");
		inPenaltyBox[currentPlayer] = true;
		
		currentPlayer++;
		if (currentPlayer == players.size()) currentPlayer = 0;
		return true;
	}


	private boolean didPlayerWin() {
		return !(purses[currentPlayer] == 6);
	}
}
